x-pack/plugin/security/cli/src/test/java/org/elasticsearch/xpack/security/cli/CertGenUtilsTests.java

```diff
@@ -100,4 +109,87 @@ public void testIsAnyLocalAddress() throws Exception {
+
+        // verify cert chaining based on PKIX rules (ex: SubjectDNs/IssuerDNs, SKIs/AKIs, BC, KU, EKU, etc)
+        final KeyStore trustStore = KeyStore.getInstance("PKCS12", "SunJSSE"); // EX: SunJSSE, BC, BC-FIPS
+        trustStore.load(null, null);
+        trustStore.setCertificateEntry("trustAnchor", rootCaCert); // anchor: any part of the chain, or issuer of last entry in chain
+
+        validateEndEntityTlsChain(trustStore, certChain, true, true);
+    }
+
+    // Wrap TrustStore in TrustManager[], cast first TrustManager to X509ExtendedTrustManager to call validation methods
+
+    /**
+     * Perform PKIX TLS certificate chain validation. This validates trust and chain correctness, not server hostname verification.
+     * Wrap TrustStore with TrustManager[], and select first element that implements the X509ExtendedTrustManager interface.
+     * Use it to call checkClientTrusted() and/or checkServerTrusted().
+     *
+     * @param trustStore       TrustStore must contain at least one trust anchor (ex: root/sub/cross/link CA certs, self-signed end entity)
+     * @param certChain        Cert chain to be validated using trust anchor(s) in TrustStore. Partial chains are supported.
+     * @param doTlsClientCheck Flag to enable TLS client end entity validation checking.
+     * @param doTlsServerCheck Flag to enable TLS server end entity validation checking.
+     * @throws Exception X509ExtendedTrustManager initialization failed, or cert chain validation failed.
+     */
+    public static void validateEndEntityTlsChain(
+        final KeyStore trustStore, final X509Certificate[] certChain,
+        final boolean doTlsClientCheck, final boolean doTlsServerCheck
+    ) throws Exception {
+        final TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance("PKIX", "SunJSSE");
+        trustManagerFactory.init(trustStore);
+        final TrustManager[] trustManagers = trustManagerFactory.getTrustManagers(); // Usually only 1 is returned
+
+        X509ExtendedTrustManager x509ExtendedTrustManager = null;
+        for (final TrustManager trustManager : trustManagers) {
+            if (trustManager instanceof X509ExtendedTrustManager) {
+                x509ExtendedTrustManager = (X509ExtendedTrustManager) trustManagers[0];
+                break; // use the first TrustManager that implements the javax.net.ssl.X509ExtendedTrustManager interface
+            }
+        }
+        if (null == x509ExtendedTrustManager) {
+            throw new UnsupportedOperationException("Expected at least one javax.net.ssl.X509ExtendedTrustManager");
+        }
+
+        // TLS authType is a substring of cipher suite. It controls what OpenJDK EndEntityChecker.java checks in KU/EKU/etc extensions.
+
+        // EKU=null|clientAuth, KU=digitalSignature, authType=DHE_DSS/DHE_RSA/ECDHE_ECDSA/ECDHE_RSA/RSA_EXPORT/UNKNOWN
+        if (doTlsClientCheck) {
```

Comment:
For my education mostly: What additional checks would `checkClientTrusted` do in this case compared to `checkServerTrusted` regarding the validity of the chain ?

Reply:
Both do same CA cert checks. Differences only apply to the first cert in the chain (i.e. the end entity cert).
- _Subject Alt Name_: Server hostname checking uses SAN dnsName & ipAddress. Client authN doesn't do anything. The caller is free to implement something else based on use case (ex: SubjectDN or SAN.directoryName for LDAP, SAN.otherName for Win DC legacy SAM account, SAN.rfc822Name for email address).
- _Key Usage_: Client authN cipher suites support RSA sign & EC sign. Server supports RSA sign, EC sign, RSA encrypt, and ECDH key agreement. Server has more flexibility.
- _Extended Key Usage_: Server supports null EKU or serverAuth. Client supports null EKU or clientAuth.

I learned by skimming through EndEntityChecker.check(). It is well documented and easy to understand. Drill down into checkTLSClient() and checkTLSServer(), and you can see the differences I mentioned for KU and EKU.
- https://github.com/AdoptOpenJDK/openjdk-jdk11/blob/master/src/java.base/share/classes/sun/security/validator/EndEntityChecker.java#L142

```
    void check(X509Certificate cert, Object parameter, boolean checkUnresolvedCritExts) throws CertificateException {
        if (variant.equals(Validator.VAR_GENERIC)) {
            return; // no checks
        }
        Set<String> exts = getCriticalExtensions(cert);
        if (variant.equals(Validator.VAR_TLS_SERVER)) {
            checkTLSServer(cert, (String)parameter, exts);                  // <===== Check this method
        } else if (variant.equals(Validator.VAR_TLS_CLIENT)) {
            checkTLSClient(cert, exts);                                     // <===== Check this method
        } else if (variant.equals(Validator.VAR_CODE_SIGNING)) {
            checkCodeSigning(cert, exts);
        } else if (variant.equals(Validator.VAR_JCE_SIGNING)) {
            checkCodeSigning(cert, exts);
        } else if (variant.equals(Validator.VAR_PLUGIN_CODE_SIGNING)) {
            checkCodeSigning(cert, exts);
        } else if (variant.equals(Validator.VAR_TSA_SERVER)) {
            checkTSAServer(cert, exts);
        } else {
            throw new CertificateException("Unknown variant: " + variant);
        }
        // if neither VAR_GENERIC variant nor unknown variant
        if (checkUnresolvedCritExts) {
            checkRemainingExtensions(exts);
        }
    }
```